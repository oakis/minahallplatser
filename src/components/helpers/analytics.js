import analytics from '@react-native-firebase/analytics';

const convertSpaces = str => str.replace(/ /g, '_');

export const track = (title, props) => {
  if (props) {
    window.log(`Firebase analytics track with props: ${title}`, props);
    analytics().logEvent(convertSpaces(title), props);
  } else {
    window.log(`Firebase analytics track: ${title}`);
    analytics().logEvent(convertSpaces(title), {});
  }
};
