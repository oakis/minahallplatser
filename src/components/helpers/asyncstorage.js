import { AsyncStorage } from 'react-native';

export const generateUid = () => {
    return `${s4()}${s4()}-${s4()}-${s4()}-${s4()}-${s4()}${s4()}${s4()}`;
};

function s4() {
    return Math.floor((1 + Math.random()) * 0x10000).toString(16).substring(1);
}

/**
 * @param {string} item 
 * @returns {Promise} Loaded storage
 */
export const getStorage = (item) => {
    return new Promise((resolve, reject) => {
        AsyncStorage.getItem(item)
        .then((json) => {
            const data = json === null ? {} : JSON.parse(json);
            window.log(`getStorage() ${item}: OK`, data);
            resolve(data);
        })
        .catch((e) => {
            window.log(`getStorage() ${item}: FAIL`, e);
            reject(e);
        });
    });
};

/**
 * @param {string} item 
 * @param {JSON} json 
 * @returns {Promise} -
 */
export const setStorage = (item, json) => {
    return new Promise((resolve, reject) => {
        AsyncStorage.setItem(item, JSON.stringify(json))
        .then(() => {
            window.log(`setStorage() ${item}: OK`);
            resolve();
        })
        .catch((e) => {
            window.log(`setStorage() ${item}: FAIL`, e);
            reject(e);
        });
    });
};
