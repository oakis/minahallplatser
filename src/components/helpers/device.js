import { Platform } from 'react-native';

export const isAndroid = () => {
    return Platform.OS === 'android';
};

export const isIOS = () => {
    return Platform.OS === 'ios';
};
