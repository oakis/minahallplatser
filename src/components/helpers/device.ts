import {Platform} from 'react-native';
import {
  getModel,
  getSystemVersion,
  getVersion,
  getUniqueId,
} from 'react-native-device-info';

export const isAndroid = (): boolean => Platform.OS === 'android';

export const isIOS = (): boolean => Platform.OS === 'ios';

export const getDeviceModel = (): string => getModel();

export const getOsVersion = (): string => getSystemVersion();

export const getAppVersion = (): string => getVersion();

export const deviceId = (): string => getUniqueId();
