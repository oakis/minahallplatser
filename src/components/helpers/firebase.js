import fetch from 'react-native-cancelable-fetch';
import { firebaseFunctionsUrl } from '../../Server';
import { handleJsonFetch } from './';

export const updateDeparturesCount = (count) => {
    if (__DEV__) {
        return;
    }
    const url = `${firebaseFunctionsUrl}/addDeparturesCount?count=${count}`;
    fetch(url, {}, 'updateDeparturesCount')
    .finally(handleJsonFetch)
    .then(({ message }) => {
        window.log(`updateDeparturesCount(): OK (${message})`);
    })
    .catch((err) => {
        window.log(`updateDeparturesCount(): FAILED' ${message})`);
        new Error(err);
    });
};

export const getDeparturesCount = () => {
    const url = `${firebaseFunctionsUrl}/getDeparturesCount`;
    return fetch(url, {}, 'getDeparturesCount')
    .finally(handleJsonFetch)
    .then(({ departuresCount }) => {
        window.log('getDeparturesCount(): OK', departuresCount);
        return departuresCount;
    })
    .catch((err) => {
        window.log('getDeparturesCount(): FAILED', err);
        new Error(err);
    });
};

export const updateStopsCount = () => {
    if (__DEV__) {
        return;
    }
    const url = `${firebaseFunctionsUrl}/addStopsCount`;
    fetch(url, {}, 'updateStopsCount')
    .finally(handleJsonFetch)
    .then(({ message }) => {
        window.log('updateStopsCount(): OK', message);
    })
    .catch((err) => {
        window.log('updateStopsCount(): FAILED', message);
        new Error(err);
    });
};

export const getStopsCount = () => {
    const url = `${firebaseFunctionsUrl}/getStopsCount`;
    return fetch(url, {}, 'getStopsCount')
    .finally(handleJsonFetch)
    .then(({ stopsCount }) => {
        window.log('getStopsCount(): OK', stopsCount);
        return stopsCount;
    })
    .catch((err) => {
        window.log('getStopsCount(): FAILED', err);
        new Error(err);
    });
};

export const incrementStopsOpened = (user, stopId) => {
    const url = `${firebaseFunctionsUrl}/incrementStopsOpen?user=${user}&stopId=${stopId}`;
    fetch(url, {}, 'incrementStopsOpened')
    .finally(handleJsonFetch)
    .then(({ message }) => {
        window.log('incrementStopsOpened(): OK', message);
    })
    .catch((err) => {
        window.log('incrementStopsOpened(): FAILED', err);
        new Error(err);
    });
}