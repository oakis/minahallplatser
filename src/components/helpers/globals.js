export const globals = {
    shouldExitApp: false,
    didLogout: false,
    isLoggingIn: true
};
