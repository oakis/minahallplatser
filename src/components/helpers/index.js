export * from './device';
export * from './message';
export * from './network';
export * from './token';
export * from './firebase';
export * from './analytics';
export * from './globals';
export * from './image';

export const isObject = (obj) => {
    return Object.prototype.toString.call(obj) === '[object Object]';
};
