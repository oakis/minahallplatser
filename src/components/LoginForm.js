import React, { Component } from 'react';
import { View, AppState, Text } from 'react-native';
import { Actions } from 'react-native-router-flux';
import { connect } from 'react-redux';
import firebase from 'react-native-firebase';
import facebook from 'react-native-fbsdk';
import { emailChanged, passwordChanged, loginUser, resetRoute, autoLogin, clearErrors, loginAnonUser, loginFacebook } from '../actions';
import { Button, Input, Message } from './common';
import { track, globals } from './helpers';
import { metrics } from './style';

const { LoginManager, AccessToken } = facebook;

class LoginForm extends Component {

	constructor(props) {
		super(props);
		this.state = {
			fbPopupVisible: false,
			fbLoading: false
		};
	}

	componentDidMount() {
		AppState.addEventListener('change', this.handleAppStateChange);
	}

	componentWillUnmount() {
		this.props.resetRoute();
		this.props.clearErrors();
		AppState.removeEventListener('change', this.handleAppStateChange);
	}

	onEmailChange = (text) => {
		this.props.emailChanged(text);
	}

	onPasswordChange = (text) => {
		this.props.passwordChanged(text);
	}

	onButtonPress = () => {
		this.props.loading = true;
		const { email, password } = this.props;
		this.props.loginUser({ email, password });
	}

	loginFacebook = () => {
		track('Login Facebook Start');
		this.setState({ fbPopupVisible: true });
		LoginManager.logInWithReadPermissions(['email'])
		.then((result) => {
			if (result.isCancelled) {
				window.log('Login cancelled:', result);
				track('Login Facebook Cancel');
			} else {
				this.setState({ fbLoading: true });
				window.log('Login success:', result);
				track('Login Facebook Success');
				AccessToken.getCurrentAccessToken().then(
					(data) => {
						globals.isLoggingIn = true;
						const credential = firebase.auth.FacebookAuthProvider.credential(data.accessToken);
						firebase.auth().signInAndRetrieveDataWithCredential(credential)
						.then(user => window.log(`Facebook account ${user.email} was successfully logged in.`))
						.catch(error => window.log('Facebook account failed:', error));
					}
				);
				this.setState({ fbPopupVisible: false });
			}
		},
		(error) => {
			window.log(`Login fail with error: ${error}`);
		})
		.catch((e) => window.log(e));
	}

	handleAppStateChange = (nextAppState) => {
		if (nextAppState === 'active' && !this.state.fbPopupVisible) {
			track('Page View', { Page: 'Login', Type: 'Reopened app from background' });
		}
	}

	render() {
		const { currentUser } = firebase.auth();
		return (
			<View
				style={{
					flex: 1,
					marginLeft: 10,
					marginRight: 10,
					flexDirection: 'column',
					justifyContent: 'center',
					alignItems: 'center'
				}}
				keyboardShouldPersistTaps="always"
				keyboardDismissMode="on-drag"
			>

				<Message
					type="danger"
					message={this.props.error}
				/>

				<Input
					placeholder="din@email.se"
					keyboardType="email-address"
					returnKeyType="next"
					onChangeText={this.onEmailChange}
					value={this.props.email}
					icon="ios-mail"
				/>
				<Input
					secureTextEntry
					placeholder="ditt lösenord"
					onChangeText={this.onPasswordChange}
					value={this.props.password}
					icon="ios-key"
					iconSize={22}
				/>

				<Button
					loading={this.props.loading}
					uppercase
					color="primary"
					label="Logga in"
					onPress={this.onButtonPress}
				/>

				<Button
					loading={this.state.fbLoading}
					color="facebook"
					uppercase
					label="Logga in med Facebook"
					onPress={this.loginFacebook}
				/>

				<Button
					fontColor="primary"
					label="Registrera"
					onPress={async () => {
						await this.props.clearErrors();
						Actions.register();
					}}
				/>
				<Button
					fontColor="primary"
					label="Glömt lösenord"
					onPress={async () => {
						await this.props.clearErrors();
						Actions.resetpw();
					}}
				/>
				{currentUser && currentUser.isAnonymous ?
					<Text
						style={{ padding: metrics.padding.md }}
						onPress={() => {
							track('Cancel Login');
							Actions.dashboard();
						}}
					>
						Gå till startsidan
					</Text> : null
				}

			</View>
		);
	}

}

const mapStateToProps = ({ auth, errors }) => {
	const { email, password, loading, token, loadingAnon } = auth;
	const { error } = errors;
	return { email, password, error, loading, token, loadingAnon };
};

export default connect(mapStateToProps, {
	emailChanged, passwordChanged, loginUser, resetRoute, autoLogin, clearErrors, loginAnonUser, loginFacebook
})(LoginForm);
