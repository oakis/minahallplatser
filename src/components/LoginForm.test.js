import React from 'react';
import { AppState } from 'react-native';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { stub } from 'sinon';
import { LoginManager } from 'react-native-fbsdk';
import firebase from 'react-native-firebase';
import LoginForm from './LoginForm';
import { track } from './helpers';

const mockStore = configureMockStore([thunk]);

const initialState = {
    auth: {
        email: 'abc@123.com',
        password: 'abc123',
        loading: false,
        token: '',
        loadingAnon: '',
    },
    errors: {
        error: 'Some login error'
    }
};

window.log = () => {};

it('should match snapshot', () => {
    const wrapper = shallow(<LoginForm store={mockStore(initialState)} />).dive();
    expect(wrapper).toMatchSnapshot();
});

describe('email input', () => {
    let wrapper;
    const emailChanged = jest.fn();
    beforeEach(() => {
        wrapper = shallow(<LoginForm store={mockStore(initialState)} emailChanged={emailChanged} />).dive();
        emailChanged.mockReset();
    });
    it('email value should match state', () => {
        const value = wrapper.find('Input').first().prop('value');
        expect(value).toBe('abc@123.com');
    });
    
    it('emailChanged should be called with the input value', () => {
        const input = wrapper.find('Input').first();
        input.simulate('changeText', 'def@123.com');
        expect(emailChanged).toBeCalledWith('def@123.com');
    });
});

describe('password input', () => {
    let wrapper;
    const passwordChanged = jest.fn();
    beforeEach(() => {
        wrapper = shallow(<LoginForm store={mockStore(initialState)} passwordChanged={passwordChanged} />).dive();
        passwordChanged.mockReset();
    });
    it('password value should match initial state', () => {
        const value = wrapper.find('Input').at(1).prop('value');
        expect(value).toBe('abc123');
    });
    
    it('passwordChanged should be called with the input value', () => {
        const input = wrapper.find('Input').at(1);
        input.simulate('changeText', 'abc456');
        expect(passwordChanged).toBeCalledWith('abc456');
    });
});

it('Logga in button should dispatch loginUser with correct props', () => {
    const loginUser = jest.fn();
    const wrapper = shallow(<LoginForm store={mockStore(initialState)} loginUser={loginUser} />).dive();
    const button = wrapper.find('Button').at(0);
    button.simulate('press');
    expect(loginUser).toBeCalledWith({ email: 'abc@123.com', password: 'abc123' });
});

describe('componentWillUnmount', () => {
    let wrapper;
    const resetRoute = stub();
    const clearErrors = stub();
    beforeEach(() => {
        wrapper = shallow(<LoginForm store={mockStore(initialState)} resetRoute={resetRoute} clearErrors={clearErrors} />).dive();
        resetRoute.reset();
        clearErrors.reset();
    });

    it('should call resetRoute', () => {
        wrapper.unmount();
        expect(resetRoute.callCount).toBe(1);
    });

    it('should call clearErrors', () => {
        wrapper.unmount();
        expect(clearErrors.callCount).toBe(1);
    });

    it('should remove event listeners', () => {
        AppState.removeEventListener = stub();
        wrapper.unmount();
        expect(AppState.removeEventListener.callCount).toBe(1);
    });
});

describe('loginFacebook', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = shallow(<LoginForm store={mockStore(initialState)} />).dive();
        track.mockReset();
    });

    it('should track start', () => {
        wrapper.instance().loginFacebook();
        expect(track).toBeCalledWith('Login Facebook Start');
    });

    it('should track cancel', async () => {
        LoginManager.logInWithReadPermissions = () => Promise.resolve({ isCancelled: true });
        await wrapper.instance().loginFacebook();
        expect(track).toBeCalledWith('Login Facebook Cancel');
    });

});
