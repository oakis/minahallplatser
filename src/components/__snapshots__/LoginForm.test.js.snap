// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <LoginForm
    email="abc@123.com"
    error="Some login error"
    loading={false}
    password="abc123"
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    storeSubscription={
      Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      }
    }
    token=""
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <Message
          message="Some login error"
          type="danger"
        />,
        <Input
          icon="ios-mail"
          keyboardType="email-address"
          onChangeText={[Function]}
          placeholder="din@email.se"
          returnKeyType="next"
          value="abc@123.com"
        />,
        <Input
          icon="ios-key"
          iconSize={22}
          onChangeText={[Function]}
          placeholder="ditt lösenord"
          secureTextEntry={true}
          value="abc123"
        />,
        <Button
          color="primary"
          label="Logga in"
          loading={false}
          onPress={[Function]}
          uppercase={true}
        />,
        <Button
          color="facebook"
          label="Logga in med Facebook"
          loading={false}
          onPress={[Function]}
          uppercase={true}
        />,
        <Button
          fontColor="primary"
          label="Registrera"
          onPress={[Function]}
        />,
        <Button
          fontColor="primary"
          label="Glömt lösenord"
          onPress={[Function]}
        />,
        null,
      ],
      "keyboardDismissMode": "on-drag",
      "keyboardShouldPersistTaps": "always",
      "style": Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "center",
        "marginLeft": 10,
        "marginRight": 10,
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "message": "Some login error",
          "type": "danger",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "icon": "ios-mail",
          "keyboardType": "email-address",
          "onChangeText": [Function],
          "placeholder": "din@email.se",
          "returnKeyType": "next",
          "value": "abc@123.com",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "icon": "ios-key",
          "iconSize": 22,
          "onChangeText": [Function],
          "placeholder": "ditt lösenord",
          "secureTextEntry": true,
          "value": "abc123",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "color": "primary",
          "label": "Logga in",
          "loading": false,
          "onPress": [Function],
          "uppercase": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "color": "facebook",
          "label": "Logga in med Facebook",
          "loading": false,
          "onPress": [Function],
          "uppercase": true,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "fontColor": "primary",
          "label": "Registrera",
          "onPress": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "fontColor": "primary",
          "label": "Glömt lösenord",
          "onPress": [Function],
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      null,
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <Message
            message="Some login error"
            type="danger"
          />,
          <Input
            icon="ios-mail"
            keyboardType="email-address"
            onChangeText={[Function]}
            placeholder="din@email.se"
            returnKeyType="next"
            value="abc@123.com"
          />,
          <Input
            icon="ios-key"
            iconSize={22}
            onChangeText={[Function]}
            placeholder="ditt lösenord"
            secureTextEntry={true}
            value="abc123"
          />,
          <Button
            color="primary"
            label="Logga in"
            loading={false}
            onPress={[Function]}
            uppercase={true}
          />,
          <Button
            color="facebook"
            label="Logga in med Facebook"
            loading={false}
            onPress={[Function]}
            uppercase={true}
          />,
          <Button
            fontColor="primary"
            label="Registrera"
            onPress={[Function]}
          />,
          <Button
            fontColor="primary"
            label="Glömt lösenord"
            onPress={[Function]}
          />,
          null,
        ],
        "keyboardDismissMode": "on-drag",
        "keyboardShouldPersistTaps": "always",
        "style": Object {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
          "marginLeft": 10,
          "marginRight": 10,
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "message": "Some login error",
            "type": "danger",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "icon": "ios-mail",
            "keyboardType": "email-address",
            "onChangeText": [Function],
            "placeholder": "din@email.se",
            "returnKeyType": "next",
            "value": "abc@123.com",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "icon": "ios-key",
            "iconSize": 22,
            "onChangeText": [Function],
            "placeholder": "ditt lösenord",
            "secureTextEntry": true,
            "value": "abc123",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "color": "primary",
            "label": "Logga in",
            "loading": false,
            "onPress": [Function],
            "uppercase": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "color": "facebook",
            "label": "Logga in med Facebook",
            "loading": false,
            "onPress": [Function],
            "uppercase": true,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "fontColor": "primary",
            "label": "Registrera",
            "onPress": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "fontColor": "primary",
            "label": "Glömt lösenord",
            "onPress": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        null,
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
